### PR TITLE
Add support direct rule

### PR DESCRIPTION
**Pull request by 5ghz**
##### SUMMARY
Add support Direct rule operations add/remove
Operation syntax
 Add Chain
```
    -  firewalld:
          chain: sshg
          fw_family: ipv4
          table: filter
          permanent: true
          state: enabled
```
Add rule to chain
```
    -  firewalld:
          chain: sshg
          fw_family: ipv4
          table: filter
          direct_rule: '-m tcp -p tcp --dport 332 -j ACCEPT'
          permanent: true
          state: enabled
```
Remove rule from chain
```
    - firewalld:
          chain: sshg
          fw_family: ipv4
          table: filter
          direct_rule: '-m tcp -p tcp --dport 332 -j ACCEPT'
          permanent: true 
          state: disabled
```
Remove chain
```
    -  firewalld:
          chain: sshg
          fw_family: ipv4
          table: filter
          permanent: true
```

Add passthrough
```
    - firewalld:
          fw_family: ipv4
          direct_rule: '-I FORWARD -o bridge3 -j ACCEPT'
          permanent: true
          state: enabled
```
Remove passthrough
```
    - firewalld:
          fw_family: ipv4
          direct_rule: '-I FORWARD -o bridge3 -j ACCEPT'
          permanent: true
          state: disabled
```

Contact: pershin87@yandex.ru
##### ISSUE TYPE
Feature Pull Request

##### COMPONENT NAME
firewalld module

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/vpershin/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

### CODE REVIEW

File: lib/ansible/modules/system/firewalld.py [hunk 1]
```diff
@@ -91,12 +88,6 @@
 notes:
   - Not tested on any Debian based system.
   - Requires the python2 bindings of firewalld, which may not be installed by default if the distribution switched to python 3
-  - Zone transactions (creating, deleting) can be performed by using only the zone and state parameters "present" or "absent".
```
Comment (maxamillion) [inline lib/ansible/modules/system/firewalld.py]: Why is this documentation being removed?

File: lib/ansible/modules/system/firewalld.py [hunk 2]
```diff
@@ -197,21 +215,19 @@ class FirewallTransaction(object):
     global module
 
     def __init__(self, fw, action_args=(), zone=None, desired_state=None,
-                 permanent=False, immediate=False, fw_offline=False,
-                 enabled_values=None, disabled_values=None):
-        # type: (firewall.client, tuple, str, bool, bool, bool)
```
Comment (maxamillion) [inline lib/ansible/modules/system/firewalld.py]: Why is the definition of `__init__` being changed? Seems out of scope for this change.

File: lib/ansible/modules/system/firewalld.py [hunk 3]
```diff
@@ -741,52 +867,6 @@ def set_disabled_permanent(self, source):
         self.update_fw_settings(fw_zone, fw_settings)
 
 
-class ZoneTransaction(FirewallTransaction):
```
Comment (maxamillion) [inline lib/ansible/modules/system/firewalld.py]: Why did you remove the ZoneTransaction functionality from the module?

File: lib/ansible/modules/system/firewalld.py [hunk 4]
```diff
@@ -799,15 +879,26 @@ def main():
             immediate=dict(type='bool', default=False),
             source=dict(required=False, default=None),
             permanent=dict(type='bool', required=False, default=None),
-            state=dict(choices=['enabled', 'disabled', 'present', 'absent'], required=True),
```
Comment (maxamillion) [inline lib/ansible/modules/system/firewalld.py]: What's the motivation for this change?

File: lib/ansible/modules/system/firewalld.py [hunk 5]
```diff
@@ -799,15 +879,26 @@ def main():
...
+            chain=dict(required=False, default=None),
+            table=dict(required=False, default=None),
+            direct_rule=dict(required=False, default=None),
+            fw_family=dict(required=False, default=None),
+            rule_priority=dict(type='int', required=False, default=0)
         ),
         supports_check_mode=True
     )
 
+    if import_failure:
```
Comment (maxamillion) [inline lib/ansible/modules/system/firewalld.py]: This should not be moved, the order of the conditionals are on purpose because the prevent from different error scenarios based on remote system state.

File: lib/ansible/modules/system/firewalld.py [hunk 6]
```diff
@@ -465,6 +501,96 @@ def set_disabled_permanent(self, service, timeout):
         self.update_fw_settings(fw_zone, fw_settings)
 
 
+class DirectRuleTransaction(FirewallTransaction):
```
Comment (maxamillion) [inline lib/ansible/modules/system/firewalld.py:505]: This looks good, would just like other comments addressed so I can understand motivation behind the proposed changes.

**Review by maxamillion — CHANGES_REQUESTED** [review]
Would like to understand the motivation for some of the changes that seem out of scope of the proposed functionality to add to the module.

Thanks.

File: lib/ansible/modules/system/firewalld.py (hunk 6 above)
Comment (5ghz) [reply lib/ansible/modules/system/firewalld.py:505]: Sorry, it was silly mistake. I used old file version.

Comment (maxamillion) [reply lib/ansible/modules/system/firewalld.py:505]: No worries, happens sometimes. :+1:

**Review by maxamillion — APPROVED** [review]

**Comment by maxamillion** [comment]
shipit

**Comment by dagwieers** [comment]
What is the status of this? People in the field need this.

**Comment by maxamillion** [comment]
Closing in favor of https://github.com/ansible/ansible/pull/49514 ... I've rebased preserving history because this should have merged a year ago and I don't expect anyone to foot the load of rebasing due to my refactor.
